dali/pipeline/operator/checkpointing/checkpoint_test.cc

```diff
@@ -0,0 +1,333 @@
+// Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/operator/checkpointing/checkpoint.h"
+
+#include <gtest/gtest.h>
+#include <stdexcept>
+#include <vector>
+
+#include "dali/test/dali_test.h"
+#include "dali/core/cuda_stream_pool.h"
+
+namespace dali {
+
+template <typename Backend>
+class DummyOperatorWithState : public Operator<Backend> {};
+
+template<>
+class DummyOperatorWithState<CPUBackend> : public Operator<CPUBackend> {
+ public:
+  explicit DummyOperatorWithState(const OpSpec &spec)
+      : Operator<CPUBackend>(spec)
+      , state_(spec.GetArgument<uint32_t>("dummy_state")) {}
+
+  void SaveState(OpCheckpoint &cpt, std::optional<cudaStream_t> stream) const override {
+    cpt.MutableCheckpointState() = state_;
+  }
+
+  void RestoreState(const OpCheckpoint &cpt) override {
+    state_ = cpt.CheckpointState<uint32_t>();
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc,
+                 const Workspace &ws) override { return false; }
+
+  uint32_t GetState() const { return state_; }
+
+ private:
+  uint32_t state_;
+};
+
+struct DummyGPUData {
+  uint32_t *ptr;
+  AccessOrder order;
+
+  inline DummyGPUData(cudaStream_t stream, uint32_t value) : order(stream) {
+    CUDA_CALL(cudaMallocAsync(&ptr, sizeof(uint32_t), stream));
+    CUDA_CALL(cudaMemcpyAsync(ptr, &value, sizeof(uint32_t),
+                              cudaMemcpyHostToDevice, stream));
+  }
+
+  inline ~DummyGPUData() {
+    CUDA_CALL(cudaFreeAsync(ptr, order.stream()));
+  }
+};
+
+template<>
+class DummyOperatorWithState<GPUBackend> : public Operator<GPUBackend> {
+ public:
+  explicit DummyOperatorWithState(const OpSpec &spec)
+      : Operator<GPUBackend>(spec)
+      , state_(spec.GetArgument<cudaStream_t>("cuda_stream"),
+               spec.GetArgument<uint32_t>("dummy_state")) {}
+
+  void SaveState(OpCheckpoint &cpt, std::optional<cudaStream_t> stream) const override {
+    if (!stream)
+      DALI_FAIL("Cuda stream was not provided for GPU operator checkpointing. ");
+    std::any &cpt_state = cpt.MutableCheckpointState();
+    cpt_state = static_cast<uint32_t>(0);
+    CUDA_CALL(cudaMemcpyAsync(&std::any_cast<uint32_t &>(cpt_state), state_.ptr,
+                              sizeof(uint32_t), cudaMemcpyDeviceToHost, *stream));
+    cpt.SetOrder(AccessOrder(*stream));
```

Comment:
In principle, if the streams were to change between the calls to SaveState, we need to set the order before scheduling the copy, so that the second copy waits for the previous one.

Reply:
Good catch!